tensornetwork/linalg/operations.py

```diff
@@ -0,0 +1,259 @@
+  out_array = tensor.backend.sqrt(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def outer(tensor1: Tensor, tensor2: Tensor) -> Tensor:
+  """Calculate the outer product of the two given Tensors."""
+  if tensor1.backend.name != tensor2.backend.name:
+    errstr = "Tried to Tensordot Tensors with differing backends \n"
+    errstr += tensor1.backend.name + "and " + tensor2.backend.name + "."
+    raise ValueError(errstr)
+  out_data = tensor1.backend.outer_product(tensor1.array, tensor2.array)
+  return Tensor(out_data, backend=tensor1.backend)
+
+
+def einsum(expression: str, *tensors: Tensor, optimize: bool) -> Tensor:
+  """Calculate sum of products of Tensors according to expression."""
+  all_backends_same, errstr = _check_backends(tensors, "einsum")
+  if not all_backends_same:
+    raise ValueError(errstr)
+  backend = tensors[0].backend
+  arrays = [tensor.array for tensor in tensors]
+  result_data = backend.einsum(expression, *arrays, optimize=optimize)
+  return Tensor(result_data, backend=backend)
+
+
+def conj(tensor: Tensor) -> Tensor:
+  """
+  Return the complex conjugate of `Tensor`
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    The complex conjugated Tensor.
+  """
+  return tensor.conj()
+
+
+def hconj(tensor: Tensor, perm: Optional[Sequence[int]] = None) -> Tensor:
+  """ The Hermitian conjugated tensor; e.g. the complex conjugate tranposed
+  by the permutation set be `axes`. By default the axes are reversed.
+  Args:
+    tensor: The Tensor to conjugate.
+    axes: The permutation. If None (default) the index order is reversed.
+  Returns:
+    The Hermitian conjugated `Tensor`.
+  """
+  return tensor.hconj(perm=perm)
+
+
+def sin(tensor: Tensor) -> Tensor:
+  """
+  Return sin of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.sin(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def cos(tensor: Tensor) -> Tensor:
+  """
+  Return cos of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.cos(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def exp(tensor: Tensor) -> Tensor:
+  """
+  Return elementwise exp of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.exp(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def log(tensor: Tensor) -> Tensor:
+  """
+  Return elementwise natural logarithm of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.log(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def diagonal(tensor: Tensor, offset: Optional[int] = 0,
+             axis1=0, axis2=1) -> Tensor:
+  """
+  Extract diagonals from a Tensor.
+
+  Args:
+    tensor: A Tensor.
+    offset     : Offset of the diagonal from the main diagonal. If tensor is 1D,
```

Comment:
typo

Reply:
Do you mean the spacing?

Reply:
Oh I see, never mind.

Reply:
yes, sorry replied to the wrong comment :)